internal/characters/arlecchino/charge.go

```diff
@@ -0,0 +1,92 @@
+package arlecchino
+
+import (
+	"github.com/genshinsim/gcsim/internal/frames"
+	"github.com/genshinsim/gcsim/pkg/core/action"
+	"github.com/genshinsim/gcsim/pkg/core/attacks"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/geometry"
+)
+
+var chargeFrames []int
+
+const chargeHitmark = 37
+
+func init() {
+	chargeFrames = frames.InitAbilSlice(60)
+	chargeFrames[action.ActionAttack] = 42
+	chargeFrames[action.ActionCharge] = 53
+	chargeFrames[action.ActionSkill] = 42
+	chargeFrames[action.ActionBurst] = 42
+	chargeFrames[action.ActionDash] = chargeHitmark
+	chargeFrames[action.ActionJump] = chargeHitmark
+	chargeFrames[action.ActionSwap] = 58
+}
+
+func (c *char) ChargeAttack(p map[string]int) (action.Info, error) {
+	windup := 0
+	if c.Core.Player.CurrentState() == action.NormalAttackState {
+		windup = 12
+	}
+
+	early, ok := p["early_cancel"]
+	if !ok {
+		early = 0
+	}
+
+	c.QueueCharTask(func() {
+		c.absorbDirectives()
+	}, 12-windup)
+
+	if early > 0 {
```

Comment:
perhaps not a require change for now but here really you should be using animation tasks. i.e. create an animation, queue all the attack tasks, then sit the animation length accordingly depends on if early cancel is called or not

take a look at klee skill

Reply:
I will open an issue for this after the PR is merged